scipy/spatial/transform/_rigid_transformation.pyx

```diff
@@ -0,0 +1,1953 @@
+        followed by p' is equivalent to `p * q`. In terms of transformation
+        matrices, the composition can be expressed as
+        ``p.as_matrix() @ q.as_matrix()``.
+
+        In terms of translations and rotations, the composition when applied to
+        a vector ``v`` is equivalent to
+        ``p.translation + p.rotation.apply(q.translation)
+        + (p.rotation * q.rotation).apply(v)``.
+
+        Parameters
+        ----------
+        other : `RigidTransformation` instance
+            Object containing the transformations to be composed with this one.
+
+        Returns
+        -------
+        `RigidTransformation` instance
+            The composed transformation. This function supports composition of
+            multiple transformations at a time. The following cases are
+            possible:
+
+            - Either ``p`` or ``q`` contains a single transformation. In this
+              case `composition` contains the result of composing each
+              transformation in the other object with the single
+              transformation.
+            - Both ``p`` and ``q`` contain ``N`` transformations. In this case
+              each transformation ``p[i]`` is composed with the corresponding
+              transformation ``q[i]`` and `output` contains ``N``
+              transformations.
+
+        Examples
+        --------
+        >>> from scipy.spatial.transform import RigidTransformation as Tf
+        >>> from scipy.spatial.transform import Rotation as R
+        >>> import numpy as np
+
+        Compose two transformations:
+
+        >>> tf1 = Tf.from_translation([1, 0, 0])
+        >>> tf2 = Tf.from_translation([0, 1, 0])
+        >>> tf = tf1 * tf2
+        >>> tf.translation
+        array([1., 1., 0.])
+        >>> tf.single
+        True
+
+        When applied to a vector, the composition of two transformations is
+        applied in right-to-left order.
+
+        >>> t1, r1 = [1, 2, 3], R.from_euler('z', 60, degrees=True)
+        >>> t2, r2 = [0, 1, 0], R.from_euler('x', 30, degrees=True)
+        >>> tf1 = Tf.from_components(t1, r1)
+        >>> tf2 = Tf.from_components(t2, r2)
+        >>> tf = tf1 * tf2
+        >>> tf.apply([1, 0, 0])
+        array([0.6339746, 3.3660254, 3.       ])
+        >>> tf1.apply(tf2.apply([1, 0, 0]))
+        array([0.6339746, 3.3660254, 3.       ])
+
+        When at least one of the transformations is not single, the result is a
+        stack of transformations.
+
+        >>> tf1 = Tf.from_translation([1, 0, 0])
+        >>> tf2 = Tf.from_translation([[0, 2, 0], [0, 0, 3]])
+        >>> tf = tf1 * tf2
+        >>> tf.translation
+        array([[1., 2., 0.],
+               [1., 0., 3.]])
+        >>> tf.single
+        False
+        >>> len(tf)
+        2
+        """
+        len_self = len(self._matrix)
+        len_other = len(other._matrix)
+        if not(len_self == 1 or len_other == 1 or len_self == len_other):
```

Comment:
If we want to be fully pedantic, then `single` attribute must participate in this check as well. Here we treat a stack of length 1 and a single transform identically.

I think both options are viable, but perhaps it's better to stick to the strict rules formulated in the docstring.

Reply:
I think the single case is covered here. The `_matrix` attribute is always stored as a `shape (N, 4, 4)` array even if single, so the `len` calls will work as expected here. And then we only enter the debug statement if both arrays are length >= 2, so the message as written for non-single transformations will always be valid. Returning a `single` result below only happens when both inputs are single.

Added a line to the docstring that two single inputs will give a single output.

Reply:
My point was whether it should work when `len(self) == 4` and `len(other) == 1` for example (both not single). Should it work?

Reply:
Ah, I see what you mean. Yes it should work, and this is the current behavior of Rotations. I suppose the more accurate analogue is that this follows the same broadcasting rules as numpy arrays. Will update the docstring to make that more clear.

Reply:
Makes sense to me.